File: 27. Delete nodes greater value on right.cpp
#include <iostream>
using namespace std; struct Node{int data; Node* next; Node(int d):data(d),next(nullptr){}
};
Node* deleteGreaterRight(Node* head){ // reverse, remove nodes with smaller value
    Node* prev=nullptr; Node* cur=head; while(cur){ Node* nxt=cur->next; cur->next=prev; prev=cur; cur=nxt; }
    Node* newHead=prev; int maxSoFar=newHead->data; Node* t=newHead; while(t && t->next){ if(t->next->data<maxSoFar){ Node* del=t->next; t->next=del->next; delete del; } else { t=t->next; maxSoFar=t->data; } }
    // reverse back
    prev=nullptr; cur=newHead; while(cur){ Node* nxt=cur->next; cur->next=prev; prev=cur; cur=nxt; }
    return prev;
}
void print(Node* h){ while(h){ cout<<h->data<<" "; h=h->next; } cout<<endl; }
int main(){ Node* head=new Node(12); head->next=new Node(15); head->next->next=new Node(10); head->next->next->next=new Node(11); head->next->next->next->next=new Node(5); head->next->next->next->next->next=new Node(6); head=deleteGreaterRight(head); print(head); return 0; }

File: 27. Delete nodes greater value on right.py
class Node:
    def __init__(self,data): self.data=data; self.next=None

def delete_greater_right(head):
    # reverse
    prev=None; cur=head
    while cur:
        nxt=cur.next; cur.next=prev; prev=cur; cur=nxt
    new_head=prev
    max_so_far=new_head.data
    t=new_head
    while t and t.next:
        if t.next.data<max_so_far:
            t.next=t.next.next
        else:
            t=t.next; max_so_far=t.data
    # reverse back
    prev=None; cur=new_head
    while cur:
        nxt=cur.next; cur.next=prev; prev=cur; cur=nxt
    return prev

def print_list(h):
    cur=h
    while cur:
        print(cur.data, end=' '); cur=cur.next
    print()

if __name__=='__main__':
    head=Node(12); head.next=Node(15); head.next.next=Node(10); head.next.next.next=Node(11); head.next.next.next.next=Node(5); head.next.next.next.next.next=Node(6)
    head=delete_greater_right(head)
    print_list(head)
